### PR TITLE
Download all missing dependencies on start-up.

### PR DESCRIPTION
**Pull request by steviegt6**
### What is the new feature?

Implements #2779.

A prompt for downloading dependencies is present as well (merged into the new/updated mods UI):
image
image
^ When no dependencies are available, it just displays the normal UI with an "OK" button.

### Why should this be part of tModLoader?

See reasoning in #2779.

### Are there alternative designs?

Needs discussion.

### CODE REVIEW

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs [hunk 1]
```diff
@@ -126,6 +126,36 @@ private enum SearchFolders { }
 			return true;
 		}
 
+		internal static IEnumerable<ulong> DownloadWorkshopDependencies() {
```
Comment (Solxanich) [inline patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs]: Can we change the method name to be 'IdentifyWorkshopDepencies'? This method does no downloading

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs [hunk 2]
```diff
@@ -126,6 +126,36 @@ private enum SearchFolders { }
...
+			HashSet<ulong> dependencies = new HashSet<ulong>();
+
+			foreach (LocalMod mod in FindWorkshopMods()) {
+				// This shouldn't really ever fail, but better safe than sorry.
+				if (!TryReadManifest(GetParentDir(mod.modFile.path), out var manifest))
```
Comment (Solxanich) [inline patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs:138]: We should consider skip dependencies checks if localMod.modreferences contains no dependencies. 

One less query to make, thus better performance.
Not required, but would be nice

File: patches/tModLoader/Terraria/ModLoader/UI/Interface.cs [hunk 3]
```diff
@@ -153,12 +155,33 @@ internal static class Interface
...
 
-					string info = ModOrganizer.DetectModChangesForInfoMessage();
-					if (info != null)
-						infoMessage.Show(Language.GetTextValue("tModLoader.ShowNewUpdatedModsInfoMessage") + info, Main.menuMode);
+					if (!skip) {
+						string newDownloads = showModChanges ? ModOrganizer.DetectModChangesForInfoMessage() : null;
+						string dependencies = promptDepDownloads ? ModOrganizer.ListDependenciesToDownload(deps) : null;
+
+						if (promptDepDownloads)
+							ModLoader.DependenciesToDownload = deps;
```
Comment (Solxanich) [inline patches/tModLoader/Terraria/ModLoader/UI/Interface.cs]: I'm not finding any reason why this is stored on ModLoader, nor needs to be. 
Remove?

**Review by Solxanich — COMMENTED** [review]
Some code review. 
A couple of minor cleanups

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs (hunk 1 above)
Comment (steviegt6) [reply patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs]: Yeah, old name from before I merged with the download/update UI.

File: patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs (hunk 2 above)
Comment (steviegt6) [reply patches/tModLoader/Terraria/ModLoader/Core/ModOrganizer.cs:138]: 👍

File: patches/tModLoader/Terraria/ModLoader/UI/Interface.cs (hunk 3 above)
Comment (steviegt6) [reply patches/tModLoader/Terraria/ModLoader/UI/Interface.cs]: Was worried about local variable capturing, but it should probably be fine in hindsight.

**Review by Chicken-Bones — APPROVED** [review]

File: patches/tModLoader/Terraria/Social/Steam/WorkshopHelper.TML.cs [hunk 4]
```diff
@@ -718,7 +719,9 @@ internal class AQueryInstance
 					ReleaseWorkshopQuery();
 				}
 
-				internal SteamUGCDetails_t FastQueryItem(ulong publishedId) {
+				// if dependencyCount > 0 then query, otherwise ignore
+				// default calls will ignore dependencies
+				internal SteamUGCDetails_t FastQueryItem(ulong publishedId, uint dependencyCount = 0) {
```
Comment (DarioDaF) [inline patches/tModLoader/Terraria/Social/Steam/WorkshopHelper.TML.cs]: Making "get dependencies" a bool and using the pDetails from the same query to get # of deps it way better

**Pull request merged**